trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/remote/RemoteS3Config.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws.proxy.server.remote;
+
+import io.airlift.configuration.Config;
+import jakarta.validation.constraints.NotNull;
+
+import java.util.Optional;
+
+public class RemoteS3Config
+{
+    private Boolean https = true;
+    private String domain = "amazonaws.com";
+    private Optional<Integer> port = Optional.empty();
+    private String virtualHostStyleTemplate = "${bucket}.s3.${region}.${domain}";
+    private String pathStyleTemplate = "s3.${region}.${domain}";
+
+    @Config("remoteS3.https")
+    public RemoteS3Config setHttps(boolean https)
+    {
+        this.https = https;
+        return this;
+    }
+
+    @Config("remoteS3.domain")
+    public RemoteS3Config setDomain(String s3Domain)
+    {
+        this.domain = s3Domain;
+        return this;
+    }
+
+    @Config("remoteS3.port")
+    public RemoteS3Config setPort(Integer port)
+    {
+        this.port = Optional.ofNullable(port);
+        return this;
+    }
+
+    @Config("remoteS3.virtualhoststyle.template")
+    public RemoteS3Config setVirtualHostStyleTemplate(String hostnameTemplate)
+    {
+        this.virtualHostStyleTemplate = hostnameTemplate;
+        return this;
+    }
+
+    @Config("remoteS3.pathstyle.template")
+    public RemoteS3Config setPathStyleTemplate(String hostnameTemplate)
```

Comment:
I'm wondering whether we need 2 settings for this. In its present form, we would need to check to make sure that these 2 are  mutually exclusive - a user should not configure both of these.

What about using a single template? That would suffice for both, with a flag to control whether we want virtual host style addressing or path style addressing.

So instead of `remoteS3.virtualhoststyle.template` and `remoteS3.pathstyle.template` we could have:
- `remoteS3.domain-template`
- `remoteS3.use-virtual-host-addressing`

Reply:
Also note my original comment yesterday relating to 2 templates was not suggesting we have a template for virtual host addressing and one for path style addressing, but rather that we can have a template for "bucketed" operations and one for "unbucketed" operations.

Currently, this implementation I believe would not work with "unbucketed" operations if using virtual host addressing.

An operation like listing buckets (which is not a "bucketed" operation) would, with the default template, resolve the hostname to `.s3.us-east-1.amazonaws.com` (us-east-1 here just being some sample region) with a leading dot, which is not valid

Reply:
However, in the present form, PathStyleRemoteS3Facade uses `remoteS3.pathstyle.template` and VirtualHostStyleRemoteS3Facade uses `remoteS3.virtualhoststyle.template`. When both properties are configured, the binding in ServerModule determines the property that will be used. If the properties are not configured, the default values are used. Though a user should configure just one of these properties, there wouldn't be a problem if both are configured.

Currently, the RemoteS3Facade that is binded is controlled here - https://github.com/trinodb/aws-proxy/blob/main/trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/TrinoAwsProxyServerModule.java#L119. We could try to leverage `remoteS3.use-virtual-host-addressing` property to bind the type of RemoteS3Facade. WDYT ?

Reply:
Added a RemoteS3Module to conditionally bind VirtualHostStyle or PathStyle.